compiler/rustc_trait_selection/src/solve/mod.rs

```diff
@@ -136,17 +142,51 @@ type CanonicalResponse<'tcx> = Canonical<'tcx, Response<'tcx>>;
 /// solver, merge the two responses again.
 pub type QueryResult<'tcx> = Result<CanonicalResponse<'tcx>, NoSolution>;
 
-pub trait TyCtxtExt<'tcx> {
+pub trait TyCtxtSolveExt<'tcx> {
     fn evaluate_goal(self, goal: CanonicalGoal<'tcx>) -> QueryResult<'tcx>;
 }
 
-impl<'tcx> TyCtxtExt<'tcx> for TyCtxt<'tcx> {
+impl<'tcx> TyCtxtSolveExt<'tcx> for TyCtxt<'tcx> {
     fn evaluate_goal(self, goal: CanonicalGoal<'tcx>) -> QueryResult<'tcx> {
         let mut search_graph = search_graph::SearchGraph::new(self);
         EvalCtxt::evaluate_canonical_goal(self, &mut search_graph, goal)
     }
 }
 
+pub(crate) trait InferCtxtSolveExt<'tcx> {
+    fn evaluate_predicates_recursively(
+        &self,
+        i: impl IntoIterator<Item = PredicateObligation<'tcx>>,
+    ) -> Result<EvaluationResult, OverflowError>;
+}
+
+impl<'tcx> InferCtxtSolveExt<'tcx> for InferCtxt<'tcx> {
+    fn evaluate_predicates_recursively(
```

Comment:
Unclear if this is the approach we want...

Alternatively, we could just canonicalize the predicates in `evaluate_predicates_recursively` and use `TyCtxtSolveExt::evaluate goal`, then instantiate the query result, then return an evaluation result...

Reply:
you can use `solve::FulfillmentCtxt` in `evaluate_predicates_recursively`

Reply:
Hm... how do we get an evaluation result out of that though?